v2/internal/util/match/string_matcher.go

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package match
+
+import (
+	"fmt"
+)
+
+// StringMatcher is an interface implemented by predicates used to test string values
+type StringMatcher interface {
+	fmt.Stringer
+
+	// Matches returns true if the provided value is matched by the matcher
+	Matches(value string) bool
+	// WasMatched returns nil if the matcher had a match, otherwise returning a diagnostic error
+	WasMatched() error
+	// IsRestrictive returns true if the matcher is populated and will restrict matches
+	IsRestrictive() bool
+}
+
+// NewStringMatcher returns a matcher for the specified string
+// Different strings may return different implementations:
+// o If the string contains '*' or '?' a globbing wildcard matcher
+// o Otherwise a case-insensitive literal string matcher
+func NewStringMatcher(matcher string) (StringMatcher, error) {
```

Comment:
An empty string should return an AlwaysMatcher

Reply:
why?

It's not used anywhere in the config today (note its removal here but CI passes), and that's not the behavior we want/need for CRD globbing - an empty string should match nothing, not everything. Otherwise it's too easy for users to accidentally include everything. They should need to be explicit and pass "*"

Reply:
I could've kept it and have split initialization that bypassed the `alwaysMatcher` possibility for the CRD globbing case but it didn't seem worth it seeing as this is never actually used in our config today. (I tested w/ a breakpoint in it, it's never hit)

Reply:
Maybe an empty string should return a `NeverMatcher`. 

My thinking was that matching an empty string using a literal matcher was brittle (programming by coincidence) and that being explicit was better. 

But maybe it's ok.

Reply:
The advantage of using `LiteralMatcher` is that:
1. `""` is a literal
2. `LiteralMatcher` has suggestions built in and so gives a better error experience than a `NeverMatcher` would (unless we implemented the same suggestions in `NeverMatcher`, I suppose).

Reply:
I can dig in a bit more to why TypeMatcher is still matching everything/working

Reply:
So it looks like, if the field is not specified at all in YAML, the `FieldMatcher` YAML deserialization override isn't even called. Actual just becomes a `nil` `StringMatcher` and we end up taking the `if dm.actual == nil {` codepath in `FieldMatcher`.

This means that there _is_ a difference between not specified and specified as `""` in the YAML. The more I think about it though, that feels like it might be a feature not a bug. Explicit "" attempting to match "" seems reasonable?